### PR TITLE
Add thread guard for `force_draw` and update related documentation

### LINKED ISSUE #79190: Calling force_draw from a thread crashes compatibility renderer
### Godot version

v4.1.stable.official.970459615

### System information

Windows 10 OpenGL API 3.3.0

### Issue description

Calling force_draw from a thread crashes compatibility renderer. Not sure if bug or this is not something that we should be doing, but it seems to not cause any issue with mobile/desktop forward+ renderer. Reason why i do this because it lets the engine run at a higher frame rate than the renderer.

### Steps to reproduce

🧵

### Minimal reproduction project

[forcedraw.zip](https://github.com/godotengine/godot/files/11988748/forcedraw.zip)

### PR DESCRIPTION
**Pull request by jsjtxietian**
Closes #79190

### CODE REVIEW

File: doc/classes/RenderingServer.xml [hunk 1]
```diff
@@ -1362,7 +1362,7 @@
 			<param index="0" name="swap_buffers" type="bool" default="true" />
 			<param index="1" name="frame_step" type="float" default="0.0" />
 			<description>
-				Forces redrawing of all viewports at once.
+				Forces redrawing of all viewports at once. Must be called from main thread.
```
Comment (akien-mga) [inline doc/classes/RenderingServer.xml:1365]:
```suggestion
				Forces redrawing of all viewports at once. Must be called from the main thread.
```

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Is documentation enough to solve #79190, which is a crash? Can we add some thread guards like we have in Node methods, or there's no access to that info from servers? CC @RandomShaper

**Comment by jsjtxietian** [comment]
I believe we should add a thread guard, should I add it to this pr or make another one ?

**Comment by akien-mga** [comment]
I think adding a thread guard in this PR too would make sense.

I see examples of it being done this way outside of Nodes (which have their own `ERR_MAIN_THREAD_GUARD`):
```
modules/navigation/godot_navigation_server_2d.cpp
186:    ERR_FAIL_COND_MSG(!Thread::is_main_thread(), "The SceneTree can only be parsed on the main thread. Call this function from the main thread or use call_deferred().");
```

Adding an error message like this would also make sense here, it could be something like:
```
ERR_FAIL_COND_MSG(!Thread::is_main_thread(), "Manually triggering the draw step from the RenderingServer can only be done on the main thread. Call this function from the main thread or use call_deferred().");
```

**Comment by RandomShaper** [comment]
I agree with @akien-mga's remarks, but I'd still need clarity on something. Is the compatibility renderer able to run on a thread (`rendering/driver/threads/thread_model == RENDER_SEPARATE_THREAD (2)`)? In that case, it'd still be possible to call any function on any thread. In that case, I'd rather put the thread guards at the `Rasterizer*GLES3` level. (For `RenderingDevice` renderers, the location would be in each `RenderingDevice*` implementation's function. After the split to `RD`-`RDD` happens, threading will be improved with guards, among other things, and I think my points here align with that plan.)

**Comment by akien-mga** [comment]
If the situation is more complex than it seemed initially, and this might partially be overridden by the RenderingDeviceDriver work, we might want to stick to just adding a note in the docs as done here for 4.2.

**Comment by RandomShaper** [comment]
Well, then let's just follow @akien-mga's idea, until everything can be addressed better.

File: servers/rendering/rendering_server_default.cpp [hunk 2]
```diff
@@ -379,6 +379,7 @@ void RenderingServerDefault::sync() {
 }
 
 void RenderingServerDefault::draw(bool p_swap_buffers, double frame_step) {
+	ERR_FAIL_COND_MSG(!::Thread::is_main_thread(), "Manually triggering the draw function from the RenderingServer can only be done on the main thread. Call this function from the main thread or use call_deferred().");
```
Comment (akien-mga) [inline servers/rendering/rendering_server_default.cpp:382]: Why the extra scope of `::Thread`?

Comment (jsjtxietian) [reply servers/rendering/rendering_server_default.cpp:382]: Just copy pasted from other places :) I will remove it.

**Review by akien-mga — APPROVED** [review]
Looks good to me, unless we foresee that this extra check could have a non-negligible performance impact.

**Comment by akien-mga** [comment]
Just in case there's a performance impact, I'm deferring this to merge after the 4.2 release, and we can cherry-pick for 4.2.x if we confirm it's a safe approach.

**Comment by YuriSizov** [comment]
Cherry-picked for 4.2.2.

**Pull request merged**